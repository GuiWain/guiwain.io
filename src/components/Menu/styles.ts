import styled, { css } from 'styled-components'

import UseLink from 'next/link'

import { Switch as UseSwitch } from 'antd'
import 'antd/lib/switch/style/index.css'

import { DarkMode, LightMode } from '@styled-icons/material/'

export const Logo = styled(UseLink)``
export const Switch = styled(UseSwitch)``

export const LightIcon = styled(LightMode)`
  ${({ theme }) => css`
    color: ${theme.colors.iconTheme};
    width: 1.6rem;
  `}
`

export const DarkIcon = styled(DarkMode)`
  ${({ theme }) => css`
    color: ${theme.colors.iconTheme};
    width: 1.6rem;
  `}
`

export const Wrapper = styled.div`
  ${({ theme }) => css`
    margin: auto;
    width: 90%;

    display: flex;
    justify-content: space-between;
    align-items: center;

    .ant-switch {
      width: 4rem;
      height: 2.2rem;
      background-color: ${theme.colors.text};

      :focus-within,
      :focus-visible {
        outline: none;
      }

      :focus {
        box-shadow: 0rem 0rem 0.2rem 0.1rem rgba(122, 0, 65, 0.2);

        .ant-switch-handle:before {
          box-shadow: inset 0rem 0rem 0.2rem 0.1rem rgba(122, 0, 65, 0.3);
        }
      }
    }

    .ant-switch-handle:before {
      background-color: ${theme.colors.primaryBg};
    }

    .ant-switch-checked .ant-switch-inner {
      margin: 0rem 2.5rem 0rem 0.2rem;
    }
  `}
`

export const Anchor = styled.a`
  ${({ theme }) => css`
    color: ${theme.colors.text};
    font-family: ${theme.font.family01};
    font-size: ${theme.font.sizes.xxlarge};
    font-weight: ${theme.font.black};

    :hover {
      color: ${theme.colors.title};
    }

    cursor: pointer;
  `}
`
