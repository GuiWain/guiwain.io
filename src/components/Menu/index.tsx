import * as S from './styles'
import { setCookie } from 'nookies'

export type Theme = 'light' | 'dark'

export type MenuProps = {
  checked: boolean
  getTheme: (theme: Theme) => void
}

const Menu = ({ checked, getTheme }: MenuProps) => {
  const handleTheme = (value: boolean) => {
    const updatedTheme = value ? 'dark' : 'light'
    const ONE_MONTH = 60 * 60 * 24 * 4

    setCookie(null, 'theme', updatedTheme, {
      maxAge: ONE_MONTH
    })

    !!getTheme && getTheme(updatedTheme)
  }

  return (
    <S.Wrapper>
      <S.Link href="/">
        <S.Anchor>GuiWain</S.Anchor>
      </S.Link>

      <S.Switch
        checkedChildren={<S.DarkIcon />}
        unCheckedChildren={<S.LightIcon />}
        onChange={(isChecked) => handleTheme(isChecked)}
        checked={checked}
      />
    </S.Wrapper>
  )
}

export default Menu
