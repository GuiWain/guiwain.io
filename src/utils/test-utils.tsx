import { ThemeProvider } from 'styled-components'
import { render, RenderResult } from '@testing-library/react'
import * as theme from 'styles/theme'

const renderWithTheme = (children: React.ReactNode): RenderResult =>
  render(<ThemeProvider theme={theme.light}>{children}</ThemeProvider>)

export * from '@testing-library/react'
export { renderWithTheme as render }
