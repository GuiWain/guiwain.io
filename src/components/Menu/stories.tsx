import { Story, Meta } from '@storybook/react'
import Menu, { MenuProps } from '.'

export default {
  title: 'Components/Menu',
  component: Menu
} as Meta

export const Default: Story<MenuProps> = (args) => <Menu {...args} />

// eslint-disable-next-line @typescript-eslint/no-empty-function
const handleTheme = () => {}

Default.args = {
  checked: true,
  getTheme: handleTheme
}
