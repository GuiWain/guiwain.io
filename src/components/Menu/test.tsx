import { render, screen, waitFor } from 'utils/test-utils'
import userEvent from '@testing-library/user-event'

import Menu from '.'

const mockArgs = {
  checked: true,
  getTheme: jest.fn()
}

describe('<Menu />', () => {
  it('should render the Menu', () => {
    const { container } = render(<Menu {...mockArgs} />)

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should invoke getTheme when click on switch', async () => {
    render(<Menu {...mockArgs} />)

    const buttonSwitch = screen.getByRole('switch')
    userEvent.click(buttonSwitch)

    await waitFor(() => {
      expect(mockArgs.getTheme).toBeCalledWith('dark')
    })
  })
})
