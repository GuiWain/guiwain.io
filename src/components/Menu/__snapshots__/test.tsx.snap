// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Menu /> should render the Menu 1`] = `
.c2 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c3 {
  color: #FFEA00;
  width: 1.6rem;
}

.c0 {
  margin: auto;
  width: 90%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 .ant-switch {
  width: 4rem;
  height: 2.2rem;
  background-color: #2D2529;
}

.c0 .ant-switch:focus-within,
.c0 .ant-switch:focus-visible {
  outline: none;
}

.c0 .ant-switch:focus {
  box-shadow: 0rem 0rem 0.2rem 0.1rem rgba(122,0,65,0.2);
}

.c0 .ant-switch:focus .ant-switch-handle:before {
  box-shadow: inset 0rem 0rem 0.2rem 0.1rem rgba(122,0,65,0.3);
}

.c0 .ant-switch-handle:before {
  background-color: #F9F9F9;
}

.c0 .ant-switch-checked .ant-switch-inner {
  margin: 0rem 2.5rem 0rem 0.2rem;
}

.c1 {
  color: #2D2529;
  font-family: Montserrat,sans-serif;
  font-size: 3.2rem;
  font-weight: 900;
  cursor: pointer;
}

.c1:hover {
  color: #7A0041;
}

<div
  class="c0"
>
  <a
    class="c1"
  >
    GuiWain
  </a>
  <button
    aria-checked="true"
    class="ant-switch ant-switch-checked"
    role="switch"
    type="button"
  >
    <div
      class="ant-switch-handle"
    />
    <span
      class="ant-switch-inner"
    >
      <svg
        aria-hidden="true"
        class="c2 c3"
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="none"
          height="24"
          width="24"
        />
        <path
          d="M12 3a9 9 0 109 9c0-.46-.04-.92-.1-1.36a5.389 5.389 0 01-4.4 2.26 5.403 5.403 0 01-3.14-9.8c-.44-.06-.9-.1-1.36-.1z"
        />
      </svg>
    </span>
  </button>
</div>
`;
